docs/book/src/developer/testing.md

```diff
@@ -11,29 +11,68 @@ Unit tests focus on individual pieces of logic - a single func - and don't requi
 be fast and great for getting the first signal on the current implementation, but unit tests have the risk of
 allowing integration bugs to slip through.
 
-Historically, in Cluster API unit tests were developed using [go test], [gomega] and the [fakeclient]; see the quick reference below.
+In Cluster API most of the unit tests are developed using [go test], [gomega] and the [fakeclient]; however using
+[fakeclient] is not suitable for all the use cases due to some limitation derived by how it is implemented,
+so in some cases contributors will be required to use [envtest]. See the [quick reference](#quick-reference) below for more details.
 
-However, considering some changes introduced in the v0.3.x releases (e.g. ObservedGeneration, Conditions), there is a common
-agreement among Cluster API maintainers that using [fakeclient] should be progressively deprecated in favor of using
-[envtest]. See the quick reference below.
+### Mocking external APIs
+In some cases when writing tests it is required to mock external API, e.g. etcd client API or the AWS sdk API.
+
+This problem is usually well scoped in core Cluster API, and in most cases it is already solved by using fake
+implementations of the target API to be injected during tests.
+
+Instead, mocking is much more relevant for infrastructure providers; in order to address the issue
+some providers can use simulators reproducing the behaviour of a real infrastructure providers (e.g CAPV);
+if this is not possible, a viable solution is to use mocks (e.g CAPA).
+
+### Generic providers
+While testing core Cluster API contributors should ensure that the code works with any providers, and thus it is required
+to not use any specific provider implementation. Instead, the so-called generic providers should be used because they
+implement the plain Cluster API contract, thus preventing developers to make assumptions we cannot rely on.
```

Comment:
```suggestion
When writing tests core Cluster API contributors should ensure that the code works with any providers, and thus it is required
to not use any specific provider implementation. Instead, the so-called generic providers e.g. "GenericInfrastructureCluster" should be used because they
implement the plain Cluster API contract. This prevents tests from relying on assumptions that may not hold true in all cases.
```

Possibly add a link to the generic providers in the code? Or at least the name of the type so they can be searched.

Reply:
I'm a little bit concerned about this because what we have is really fragmented. Even if not ideal, I prefer contributors to ask for guidance on this point so we can weigh pros and cons case by case; also, using the v1.1 timeframe as a benchmark, changes of this type are not really frequent.

Reply:
I think this is fine, given that a link might be more confusing. Adding the name of the example type, i.e. GenericInfrastructureCluster,  lets people search for it themselves in the codebase to see the implementation